rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,393 @@
+given.
+
+This does not implement or intend to implement content-negotiation, instead is
+a system to ensure that only requests which would've negotiated a compatible
+encoding would succeed.
+
+## API
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  These requests are used
+to transmit metadata of an HTTP request which the service requires the client
+to perform.
+
+An example usage of this format is uploading a file.  The creation endpoint of
+this service will generate all of the HTTP request metadata that the uploading
+process will use to upload file.  Then the uploader will match these HTTP
+requests with payload bodies and run each of these requests.
+
+The motivation for this type of flow is to support the case where the uploader
+is running in a different region than this service is.  If we required all
+uploads to gate through this service, we'd need to duplicate uploading effort
+and perform unneeded inter-region transfers.  By transmitting a list of
+requests which allow the uploading to happen, we can instruct the uploader to
+upload directly to the existing S3 infrastructure, but perform all signing
+within this service.
+
+A generalised request will look like:
+
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in
+  the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key ->
+  list of strings)
+* All query string, path, fragments are contained in a fully formed `url`
+  property
+
+### Origins
+Where present, the `origin` rest parameter in this service will be either an
+IPv4 address, IPv6 address, or an identifier.  This parameter will specify the
+source location of the ultimately originating request to upload.  If the origin
+is an IPv4 or IPv6, a mapping of the address and IP will occur to find the
+backing storage.  If the origin is not an IPv4 or IPv6 address and is identical
+to a set of known and mappable identifiers (e.g.  `s3_us-west-2`), then that
+will be used to map the token to find the backing storage provider.  If the
+origin is not an IP address and not a configured identifier, an error will be
+sent back to the client.
+
+The mapping of origins to storage providers should be statically configured.
+
+The origin `all` is reserved and cannot be used.  This is to enable management
+endpoints to operate on all origins in the future.
+
+### Creation
+
+```
+PUT /objects/:name
+PATCH /objects/:name
+```
+
+The `PUT` endpoint will take an origin and name.  It will return a `200` with a
+document that contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the object in the shape:
```

Comment:
How are errors handled?

Reply:
by responding with 400 and 500 series responses as appropriate.